src/Parser.cpp

```diff
@@ -158,20 +155,23 @@ void PacketParser::process_packets(err_code_t& err) {
             mt_kvPtr->cas_unique = 0;
           } else {  // gets
             READ_UNSIGNED(mt_kvPtr->cas_unique);
-            SKIP_BYTES(1); // '\r' after cas
+            SKIP_BYTES(1);  // '\r' after cas
           }
+          m_state = FSM_GET_CAS;
+        }
+        break;
+      case FSM_GET_CAS: // got "cas\r"
+        {
+          assert(mt_kvPtr != NULL && mt_kvPtr->bytesRemain == mt_kvPtr->bytes + 1);
+          SKIP_BYTES(1);  // '\n'
+          --mt_kvPtr->bytesRemain;
           m_state = FSM_GET_VALUE_REMAINING;
         }
         break;
-      case FSM_GET_VALUE_REMAINING: // not got "\n" + all bytes + "\r\n"
+      case FSM_GET_VALUE_REMAINING: // not got <data block> + "\r\n"
         {
-          assert(mt_kvPtr != NULL);
-          if (mt_kvPtr->bytesRemain == mt_kvPtr->bytes + 1) {
-            SKIP_BYTES(1);
-            --mt_kvPtr->bytesRemain;
-          }
-
-          if (mt_kvPtr->bytesRemain == mt_kvPtr->bytes) {
+          assert(mt_kvPtr != NULL && (mt_kvPtr->bytesRemain == mt_kvPtr->bytes || mt_kvPtr->bytesRemain == 0));
```

Comment:
I don't think `mt_kvPtr->bytesRemain == mt_kvPtr->bytes || mt_kvPtr->bytesRemain == 0`  is alway true. If the whole buffer is splitted into chunks and at one time the client may only receive one chunk, which means `0 < bytesRemain < bytes`. Please correct if I'm wrong.

Reply:
@mckelvin this is a interesting point，https://github.com/douban/libmc/pull/68/files#diff-4ddea0926f3b6d0363f3f47725b88958R179 m_buffer_reader->readBytes want to read "mt_kvPtr->bytes" bytes，https://github.com/douban/libmc/blob/624ca7322a1248e7b81cee7028bde2b98585bba6/src/BufferReader.cpp#L340 if it cant, seems like it's RET_INCOMPLETE_BUFFER_ERR，then just return。will try next time。if it sucess，"mt_kvPtr->bytesRemain = 0; "

so i think mt_kvPtr->bytesRemain is 0 or mt_kvPtr->bytes。

Reply:
@mckelvin i say it's interesting，because maybe the m_buffer_reader->readBytes's behavior can improve，but it also means this line https://github.com/douban/libmc/blob/master/src/BufferReader.cpp#L178 's behavior need to adjust，because this line seems like assume you will not peek the current m_blockWriteIterator。

Reply:
You're right. 👍